datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/KafkaPositionTracker.java

```diff
@@ -541,20 +573,79 @@ protected void runOneIteration() {
         }
       }));
 
-      // Query the broker for its offsets for those partitions
       try {
-        _kafkaPositionTracker.queryBrokerForLatestOffsets(_consumer, partitionsNeedingUpdate);
+        if (_enablePartitionLeadershipCalculation) {
+          updatePartitionLeadershipMap();
+        }
       } catch (Exception e) {
-        LOG.warn("Failed to query latest broker offsets via endOffsets() RPC", e);
-        throw e;
+        LOG.warn("Failed to update the partition leadership map. Using stale leadership data. Reason: {}", e.getMessage());
+        LOG.debug("Failed to update the partition leadership map. Using stale leadership data.", e);
+      }
+
+      // Query the broker for its offsets for those partitions
+      batchPartitionsByLeader(partitionsNeedingUpdate).forEach(partitionLeaderBatch -> {
+        try {
+          _kafkaPositionTracker.queryBrokerForLatestOffsets(_consumer, partitionLeaderBatch, _consumerRequestTimeout);
+        } catch (Exception e) {
+          LOG.warn("Failed to query latest broker offsets for this leader batch via endOffsets() RPC. Reason: {}", e.getMessage());
+          LOG.debug("Failed to query latest broker offsets for this leader batch via endOffsets() RPC", e);
+        }
+      });
+    }
+
+    /**
+     * Queries a Kafka broker for the leader of each partition and caches that information.
+     */
+    private void updatePartitionLeadershipMap() {
+      Optional.ofNullable(_consumer.listTopics(_consumerRequestTimeout)).ifPresent(response -> {
+        Map<TopicPartition, Node> updateMap = response.values().stream()
+            .filter(Objects::nonNull)
+            .flatMap(Collection::stream)
+            .filter(partitionInfo -> partitionInfo != null && partitionInfo.topic() != null && partitionInfo.leader() != null)
+            .collect(Collectors.toMap(
+                partitionInfo -> new TopicPartition(partitionInfo.topic(), partitionInfo.partition()),
+                PartitionInfo::leader));
+        _partitionLeadershipMap.keySet().retainAll(updateMap.keySet());
+        _partitionLeadershipMap.putAll(updateMap);
+      });
+    }
+
+    /**
+     * Groups topic partitions into batches based on their leader.
+     * @param topicPartitions the topic partitions to group by leader
+     * @return a list of topic partitions batches
+     */
+    @NotNull
+    private List<Set<TopicPartition>> batchPartitionsByLeader(@NotNull Set<TopicPartition> topicPartitions) {
+      if (!_enablePartitionLeadershipCalculation) {
```

Comment:
nit: might want to do it if _partitionLeadershipMap is empty/null

Reply:
_partitionLeadershipMap will never be null in this path, but I will add the empty check.